xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -184,45 +186,53 @@ public void handleSubchannelState(Subchannel subchannel, ConnectivityStateInfo n
     public void reset() {
       for (XdsLocality locality : localityMap.keySet()) {
         localityMap.get(locality).shutdown();
-        loadStatsStore.removeLocality(locality);
       }
       localityMap = ImmutableMap.of();
+      for (XdsLocality locality : edsResponsLocalityInfo.keySet()) {
+        loadStatsStore.removeLocality(locality);
+      }
     }
 
     // This is triggered by EDS response.
     @Override
     public void updateLocalityStore(Map<XdsLocality, LocalityInfo> localityInfoMap) {
       Set<XdsLocality> oldLocalities = localityMap.keySet();
       Set<XdsLocality> newLocalities = localityInfoMap.keySet();
-      Map<XdsLocality, LocalityLbInfo> updatedLocalityMap = new LinkedHashMap<>();
+      ImmutableMap.Builder<XdsLocality, LocalityLbInfo> updatedLocalityMap = ImmutableMap.builder();
 
-      final Set<XdsLocality> toRemove = new HashSet<>();
       for (XdsLocality oldLocality : oldLocalities) {
-        if (!newLocalities.contains(oldLocality)) {
-          toRemove.add(oldLocality);
-          // No graceful transition until a high-level lb graceful transition design is available.
-          localityMap.get(oldLocality).shutdown();
+        if (!newLocalities.contains(oldLocality)
+            && !localityMap.get(oldLocality).isDeactivated()) {
```

Comment:
The check for `isDeactivated()` is unnecessary because `deactivate()` also does it.

Reply:
Done.

Reply:
Actually I think checking `isDeactiveated()`is necessary here and the one in `deactivate()` should be removed. Here we are making the decision to deactivate or not, but the check in `deactivate()` is just for checking the precondition of the method call. Weakening the precondition isn't always good as we are hiding away potential errors.

Reply:
I think it's not necessary `if (!newLocalities.contains(oldLocality)` then deactivate `oldLocality` is very natural, and if it's already deactivate, the no-op is also natural. Adding precondition error check increases complexity.

Reply:
I am not saying adding precondition check. I am saying you should remove the precondition check as it is hiding potential errors.

Where to put the check depends on what's the spec for `deactivate()` method. If it is
> Schedule a timer to be fired after 15 minutes to remove the given locality from `localityMap`. Modifies `localityMap`. Requires the given locality in `localityMap`.

Then you would want to do it before (outside) the method call to ensure the precondition.

If your precondition is
> Schedule a timer to be fired after 15 minutes to remove the given locality from `localityMap`. Do thing if the given locality does not exist in `localityMap`. Modifies `localityMap`.

Then you would want to do it in `deactivate()`.

Putting the check on both places blurs the logic and hides potential errors.